services/QuillLMS/client/app/bundles/Teacher/startup/CmsDistrictIndexAppClient.tsx

```diff
@@ -0,0 +1,236 @@
+      newState.query.sort_direction = sort_direction
+      this.setState(newState, this.search)
+    }
+  };
+
+  search = (e) => {
+    const { query } = this.state
+
+    e ? e.preventDefault() : null
+    this.setState({loading: true})
+    const link = `${process.env.DEFAULT_URL}/cms/districts/search`
+    const data = new FormData();
+    data.append('district', '')
+    Object.keys(query).forEach((k) => {
+      data.append(k, query[k])
+    })
+    fetch(link, {
+      method: 'POST',
+      mode: 'cors',
+      credentials: 'include',
+      body: data,
+      headers: {
+        'X-CSRF-Token': getAuthToken()
+      }
+    }).then((response) => {
+      if (!response.ok) {
+        throw Error(response.statusText);
+      }
+      return response.json();
+    }).then((response) => {
+      this.setState({ data: response.districtSearchQueryResults, numberOfPages: response.numberOfPages, loading: false })
+    }).catch((error) => {
+      // to do, use Sentry to capture error
+    })
+  };
+
+  submitPageForm = e => {
+    this.updatePage(e.target.page.value)
+  };
+
+  updateField(e, key) {
+    const value = e.target.value
+    const newState = { ...this.state}
+    newState.query[key] = value
+    this.setState(newState)
+  }
+
+  updatePage = i => {
+    const newState = { ...this.state}
+    newState.query.page = i
+    this.setState(newState, this.search)
+  };
+
+  renderPageSelector() {
+    const { query, numberOfPages } = this.state
+
+    const currentPage = query.page || 1
+    const totalPages = numberOfPages || 1
+    return (
+      <div style={{display: 'flex', justifyContent: 'space-between', alignItems: 'center'}}>
+        <a onClick={() => this.updatePage(1)}>First</a>
+        <form onSubmit={this.submitPageForm}>
+          <input defaultValue={currentPage} name='page' /><span>of {totalPages}</span>
+        </form>
+        <a onClick={() => this.updatePage(totalPages)}>Last</a>
+      </div>
+    )
+  }
+
+  renderTableOrLoading() {
+    const { loading, data, columns, query } = this.state
+
+    if (loading) {
+      return <LoadingIndicator />
+    } else if (data && data.length) {
+      const sort = query.sort ? query.sort : 'number_teachers'
+      const sortDescending = query.sort_direction ? query.sort_direction === 'desc' : true
+      return (
+        <div>
+          <ReactTable
+            className='progress-report activity-scores-table'
+            columns={columns}
+            data={data}
+            defaultPageSize={100}
+            defaultSorted={[{id: sort, desc: sortDescending}]}
+            minRows={1}
+            onSortedChange={this.setSort}
+            showPageSizeOptions={false}
+            showPagination={false}
+            showPaginationBottom={false}
+            showPaginationTop={false}
+          />
+          <div className='cms-pagination-container'>
+            {this.renderPageSelector()}
+          </div>
+        </div>
+      )
+    } else {
```

Comment:
Take or leave, but consider moving this clause to the top to have less branching, e.g. a few "guard" clauses, then the code.

```javascript
if (loading) { return <LoadingIndicator />}

if (!data || data.length === 0) { return <p>No records found for your query.</p> }

// The rest of the happy path code with no branching.